File: 49-group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagrams = new HashMap<>();

        for (String s : strs) {
            char[] chars = s.toCharArray();
            Arrays.sort(chars);
            String sorted = new String(chars);

            anagrams.computeIfAbsent(sorted, k -> new ArrayList<>()).add(s);
            // if (!anagrams.containsKey(sorted)) {
            //     anagrams.put(sorted, new ArrayList<>());
            // }
            // anagrams.get(sorted).add(s);
        }

        return new ArrayList<>(anagrams.values());
    }
}

File: 49-group-anagrams/group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # key -> [anagrams]
        anagrams = collections.defaultdict(list)

        for s in strs:
            key = [0] * 26
            for c in s:
                key[ord(c) - ord('a')] += 1
            anagrams[tuple(key)].append(s)
        
        return anagrams.values()
            

